### DISCUSSION TITLE
Handling of RAR files

### DISCUSSION THREAD

**Discussion by MrZigi** [mikekazakov/nimble-commander#409] [Q&A] [answered]
Hello Mike,

I'm a bit behind with NC updates so have a question if you could spare a minute or so.
I see this item in the release note of 1.5: "Improved the handling of RAR archives".
 
1. As a devoted long-time user of RAR where I could find some more documentation about this feature? 
2. I see that e.g. files can be extracted on the fly while using the Copy or Preview function (not sure if available earlier). That's already great! Any chance this works once with password protected RAR archives?
3. Was there anything added to the NC interface to support compress w/ RAR? I'm using it as a defined Command Tool (not so perfect) and I'm still dreaming of seamless integration like offered by Compress function.   

As usual, big thanks to the excellent job with NC! It's a great everyday aid to my work.
Thanks & best regards,
Zigi

**Comment by mikekazakov** [comment, marked as answer]
Hey, sure:

> 1. As a devoted long-time user of RAR where I could find some more documentation about this feature?

Before v1.5, Nimble Commander used the [UnRAR](https://www.rarlab.com/rar_add.htm) library to deal with RAR archives, it was a separate VFS, written for this type specifically. It came with a swath of problems, starting with the poor integration with NC itself and ending with missing some of the format options. Starting with v1.5 RAR archives are handled using the [libarchive](https://github.com/libarchive/libarchive) library, as any other archive types.

> 2. I see that e.g. files can be extracted on the fly while using the Copy or Preview function (not sure if available earlier). That's already great! Any chance this works once with password protected RAR archives?

No, this won't work - libarchive does not support reading password-protected RAR archives.

> 3. Was there anything added to the NC interface to support compress w/ RAR? I'm using it as a defined Command Tool (not so perfect) and I'm still dreaming of seamless integration like offered by Compress function.

Currently the only supported format for compression is ZIP. While other formats can be added in theory, RAR is currently out of the question - libarchive does not implement compression for this format.

> As usual, big thanks to the excellent job with NC! It's a great everyday aid to my work.

Thanks, you're welcome!

**Comment by MrZigi** [comment]
Hello Mike and to whom it may concern,

finally found the time to figure out why some RAR archives open in NC (and Finder) and others not. 
It turns out solid archives (option -s) do not open. That's pity because solid archives are significantly smaller (sometimes as much as 50%). 

Thanks & regards,
Zigi

**Reply by mikekazakov** [reply]
Hm, it looks like solid rar archives should be supported by the library NC uses under the hood, i.e. libarchive: https://github.com/libarchive/libarchive/commit/087319f7060842233e0a705654d855e77ec71ff7

Could you check if you can extract the archive contents using `bsdtar` which is built directly on top of libarchive?

**Reply by MrZigi** [reply]
here is the result:
bsdtar: RAR solid archive support unavailable.
PS.
Just in case it matters: running Catalina 10.15.7.

**Reply by mikekazakov** [reply]
Well, it explains then - first libarchive needs to properly support these types of archives.

**Reply by MrZigi** [reply]
Please excuse my ignorance Mike but could you explain it a bit more?
Initially I thought this libarchive is a part of macOS but it seems to be open source project. 

1. So how did it make into my system if not with NC? 
2. If with NC and I'm using 1.6.1 which is up-to-date so how come libarchive being potentially outdated and not supporting solid archives?
3. How I can check libarchive version in my system?

Thanks & regards,
Zigi

**Reply by mikekazakov** [reply]
Sure, libarchive originated in FreeBSD, Apple just incorporates this library: https://en.wikipedia.org/wiki/Libarchive.
1. `bsdtar` and other parts of libarchive are shipped with macOS itself. NC builds and ships internally its own version of libarchive independently of macOS it runs on.
2. Well this functionality might be simply not implemented in the library yet. It's an OSS project and the RAR format is not the most popular one.
3. `bsdtar --version` will give you this information. macOS 15.1 reports `3.5.3`, NC v1.7 uses `3.7.4`: https://github.com/mikekazakov/nimble-commander/blob/002ef66bad06359f7da3838f782d5e88dd1bb617/3rd_Party/versions.txt#L10

**Reply by MrZigi** [reply]
NC 1.7? I can only see 1.6.0 in the App store and 1.6.1 as direct download as well as release info here.

**Reply by mikekazakov** [reply]
1.7 is currently in development and is available in nightly builds. 1.6 uses the same version of libarchive.

**Reply by MrZigi** [reply]
If 1.6 uses 3.7.4 of libarchive then it should not depend of the version of my macOS, correct? And in such case open solid archives, right?

**Reply by mikekazakov** [reply]
Yes, NC does not depend on the version of `libarchive` macOS ships.
To check if `libarchive` actually supports solid RAR archives (and not just says so in the code) - you need to get a fresh version of `bsdtar` and try that one. e.g. from Brew: `https://formulae.brew.sh/formula/libarchive`.
